Use Post Type label for Document tab in Settings Header

desaiuditd opened the pull request:
## Description
This PR initially started as ability to filter the Document label in the sidebar. I needed to be able to rename that label in my project. So that I could differentiate between different post types and custom post types as well. It could also be a generic custom label e.g., `Article`, `Story` etc. for any specific newsroom.

![image](https://user-images.githubusercontent.com/2236554/64170729-2a335680-ce94-11e9-8a86-9b4f877c20f7.png)

After many iterations, eventually we decided to use Post Type label for this. So that we can also help in #21850. For my use case, I will be able to filter the post type labels probably using an existing WP filter `register_post_type_args` or maybe an existing WP action right before the Block Editor preloads initial data `block_editor_preload_paths`.

## How has this been tested?
Tested this on my local WordPress instance by checking Template and Template Parts CPTs pages.

## Screenshots
Result after adding some custom code for my use case:
![image](https://user-images.githubusercontent.com/2236554/64171068-0de3e980-ce95-11e9-83bc-c958aab2d562.png)

Result on the Template and Template Parts CPTs registered for FSE.
![image](https://user-images.githubusercontent.com/2236554/83327087-a7251c80-a2bc-11ea-9ab4-5c826569089c.png)

![image](https://user-images.githubusercontent.com/2236554/83327110-d0de4380-a2bc-11ea-8d28-bfff4f66f4c8.png)


## Types of changes
- JS changes to read the label value from Post Type label rather than `Document`. If the post type does not have the label value, then it will fallback to `Document`. This is a behaviour change.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

desaiuditd committed:
Introduce a filter to overrider Document label in Document Settings Header.

jorgefilipecosta commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -11,11 +12,12 @@ import SidebarHeader from '../sidebar-header';
 	const [ documentAriaLabel, documentActiveClass ] = sidebarName === 'edit-post/document' ?
 		// translators: ARIA label for the Document sidebar tab, selected.
-		[ __( 'Document (selected)' ), 'is-active' ] :
+		[ `${ documentLabel } ${ __( '(selected)' ) }`, 'is-active' ] :
```
Normally we try to avoid concatenation of translatable strings. In another language "(selected)" may need to be before the document label.
What if we use an sprintf?
```
/* translators: 1: document label */
sprintf( __( '%1$s (selected)' ) )
```

swissspidy replied:
`sprintf` is definitely preferred over string concatenation.

desaiuditd replied:
c2d4dc0

jorgefilipecosta commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -11,11 +12,12 @@ import SidebarHeader from '../sidebar-header';
 
 const SettingsHeader = ( { openDocumentSettings, openBlockSettings, sidebarName } ) => {
 	const blockLabel = __( 'Block' );
+	const documentLabel = applyFilters( 'editor.sidebar.settingsHeader.documentLabel', __( 'Document' ) );
```
Hi @gziolo, any thoughts related to this new filter?

desaiuditd replied:
I could also use some help in the naming of this filter, if this is the right approach moving forward.

gziolo replied:
I shared my thoughts here: https://github.com/WordPress/gutenberg/pull/17311#pullrequestreview-283376728. I marked PR as `request changes` only to ensure that we have a proper discussion about customizing UI labels with some guidelines about implementation documented before we land this change.

aprea commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -11,11 +12,12 @@ import SidebarHeader from '../sidebar-header';
 		// translators: ARIA label for the Document sidebar tab, selected.
-		[ __( 'Document (selected)' ), 'is-active' ] :
+		[ `${ documentLabel } ${ __( '(selected)' ) }`, 'is-active' ] :
 		// translators: ARIA label for the Document sidebar tab, not selected.
```
This comment may be better situated above line 15.

desaiuditd replied:
0de2bb4

desaiuditd committed:
Move the translator comment for ARIA label.

desaiuditd committed:
Replace string concatenation with sprintf.

desaiuditd commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -11,11 +12,12 @@ import SidebarHeader from '../sidebar-header';
 const SettingsHeader = ( { openDocumentSettings, openBlockSettings, sidebarName } ) => {
 	const blockLabel = __( 'Block' );
+	// translators: ARIA label for the Document sidebar tab, not selected.
+	const documentLabel = applyFilters( 'editor.sidebar.settingsHeader.documentLabel', __( 'Document' ) );
```
Also, I did not see any existing pattern to document `applyFilters`.

E.g., in PHP, we document the `apply_filters` wherever they are introduced along with the parameters and return values.

Do we have such pattern to add documentation for JS filters/hooks?

[Block Filters](https://developer.wordpress.org/block-editor/developers/filters/block-filters/) page does not seem like a good place for these extra utility hooks, as they are not directly related to blocks.

ffxluca reviewed (approved):
👏

gziolo reviewed (dismissed):
Thank you for opening this pull request.

To be clear. I understand the need to have more ways to customize the user interface. At the same time, I would like to notice that we don't use JS filters for UI labels in any place in the block editor. There are a few labels which are filterable using PHP filters, like the Paragraph block placeholder which is passed to the editor through its settings:

https://github.com/WordPress/gutenberg/blob/7964f381e07e06b2ac13698fd1465a4d0c16da1b/packages/block-editor/src/components/default-block-appender/index.js#L82

Related PHP filter:
https://github.com/WordPress/wordpress-develop/blob/00b03f2a6f2c21ebf9975c9c08976358681400ba/src/wp-admin/edit-form-blocks.php#L270

This should be discussed first, what's the best strategy moving forward with adding new filters for UI labels and what naming conventions should be used before we proceed with this particular use case.

desaiuditd commented:
I also thought this to be an editor level setting first, but I felt it's not affecting the whole editor and only limited to the sidebar section. Also, I could not find any single place where all the labels are listed as a source of truth. So ended up adding the filter in the component itself.

But if this is useful in wider sense, then taking a generic approach and a single collection of all the labels at one place makes more sense.

My next question would be, which labels are good candidates to be part of this collection? Not everyone would like to customise them. So should it be part of WordPress Core or the Gutenberg plugin itself?

aprea commented:
>This should be discussed first, what's the best strategy moving forward with adding new filters for UI labels and what naming conventions should be used before we proceed with this particular use case.

@gziolo where is the best place to kick off this discussion? In this PR?, the #core-editor channel, elsewhere?

swissspidy commented:
@aprea @gziolo We could discuss this in tomorrow's core editor meeting on Slack

swissspidy commented:
Reminds me a bit of #12517 by the way :-)

desaiuditd commented:
> Reminds me a bit of #12517 by the way :-)

Wow, that's wild. I would not have imagined that. 😃

gziolo commented:
> We could discuss this in tomorrow's core editor meeting on Slack

Sure, it sounds like a great topic to discuss.

> Reminds me a bit of #12517 by the way :-)

I was thinking about it the other day. It's definitely one of the options as long as we ensure it performs well enough so you don't have to apply the same filters all over again on every component's re-render. 

I was also thinking about the alternatives we could investigate. I came up with one idea so far. Introduce `useTranslation` React hook which you would have to explicitly call with some unique name for the filter to open a given string for modifications. The rationale behind it is, that sometimes you use the same translation string in many places and you don't necessarily want to update all of them at the same time. Having React hook gives us also more flexibility as we could re-render component when a new filter is applied for the translation, even after the page was rendered. 

Well, we could also discuss some set of constraints which would make the initial implementation safer if we do something like #12517. Example:
- all filters need to be applied before the initial render
- only translations with context applied can be filterable

desaiuditd commented:
Thoughts on this from last Weekly Editor Chat:

@gziolo: 
> We still would want to use applyFilters, yes
We were concerned in the past about performance implications and now I think that another issues is the discovery of all filterable translations

@youknowriad:
> One "somehow related" idea I was thinking about is the fact that applyFilters is very sensitive to the scripts loading order and we have an opportunity to build a useFilter alternative to fix this.

desaiuditd commented:
@youknowriad, if I'm to take your approach of introducing new `useFilters` hook, in which package, would you put that?

youknowriad commented:
If we were to add a `useFilters` it should probably go in the `filters` package. I guess this means adding a `@wordpress/element` dependency to that package but maybe it's ok? Thoughts?

gziolo commented:
> If we were to add a `useFilter` it should probably go in the `filters` package. I guess this means adding a `@wordpress/element` dependency to that package but maybe it's ok? Thoughts?

I was thinking, we might want to place it inside `@wordpress/plugins` which depends on both `@wordpress/element` and `@wordpress/hooks` already.

desaiuditd commented:
Oh. A new package of its own?

> adding a @wordpress/element dependency

This should be fine, I guess.

What qualifies for a new package? What purpose it's going to serve? What else it's going to contain?

> @wordpress/plugins
This seems fine. I would not have much of an opinion on this (Lack of knowledge). Would defer to others.

youknowriad commented:
I was actually thinking about the `hooks` package (my brain said `filters`) :)

gziolo commented:
> I was actually thinking about the `hooks` package (my brain said `filters`) :)

It should work as well. Having `useFilter` hook inside `@wordpress/hooks` might be an interesting challenge in terms of explaining the difference between WordPress and React hooks :) However, this is what it is regardless. Just an observation. I think @youknowriad is right.

@adamsilverstein - any thoughts on this topic as you are expert in terms of WordPress hooks :)

desaiuditd commented:
`@wordpress/hooks` seems to be a plain JavaScript package. Also, there's a general vibe that `@wordpress/hooks` can be used in any other system which may not necessarily use React. So do we have to maintain that backward compatibility?

adamsilverstein commented:
I'm not sure I understand fully how `useFilters` would work, sounds interesting - could you give some pseudo code to explain how you would use it? 

the `wp.hooks` package was designed to work in plugins and replicate the plugin php api, i'm not sure adding a react pattern dependency makes sense - especially given the already confusing naming.

peterwilsoncc commented:
> the `wp.hooks` package was designed to work in plugins and replicate the plugin php api, i'm not sure adding a `React` pattern dependency makes sense - especially given the already confusing naming.

`wp.hooks` seems more sensible to me rather than react approach, too. The wheelhouse is much closer to the modifications/setting changes currently taking advantage of the WP approach.

ellatrix commented:
Also not use how `useFilters` would work, since you need to wrap an existing component around a new component. I think you _need_ to create a higher order component for this, there's no other way around it?

We do have some use cases of `withFilters` where we're only filtering props (`RichText` comes to mind). For this case, we could create a `useFilters` hooks to only filters the props. Or perhaps we should try to use the context API for it.

desaiuditd commented:
I was thinking about this last week. And that's the exact problem I faced while working on a prototype. To be able to filter the react component, I'll need to re-render them or at least force update them. Similar to `withFilters`.

At the end, I concluded a middle ground. How about we introduce `useFilters` only for primitive types of JavaScript like `Number`, `String`, `Boolean`, `Object`, `Array`, `null`, `undefined` etc.

And for React components, we already have `withFilters` working well.

Main reason to reach this proposal was also the fact that I did not want to re-engineer what is already working in `withFilters`.

Does that work? Thoughts?

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

# Conflicts:
#	packages/edit-post/src/components/sidebar/settings-header/index.js

desaiuditd committed:
Introduce useFilters React hook.

desaiuditd commented:
@youknowriad @gziolo I've pushed a new commit 8652601 to introduce `useFilters`.

The location of this hook can be questionable (currently put inside `@wordpress/components` in parallel to `withFilters`). We can debate on that.

But I wanted to get some early feedback on the approach. As mentioned in my previous comment,

> `useFilters` only supports primitive types of JavaScript like `Number`, `String`, `Boolean`, `Object`, `Array`, `null`, `undefined` etc.
>
> And for React components, we already have withFilters working well.
>
> Main reason to reach this proposal was also the fact that I did not want to re-engineer what is already working in withFilters.

Can you please review this? I can polish this PR further once I get an OK on the approach.

ZebulanStanphill commented on packages/components/src/higher-order/use-filters/index.js:
```diff
@@ -0,0 +1,40 @@
+import { addAction, applyFilters, removeAction } from '@wordpress/hooks';
+
+const useFilters = ( hookName, originalValue, ...args ) => {
+	const namespace = 'core/use-filter/' + hookName;
```
The [Block Editor Coding Guidelines](https://developer.wordpress.org/block-editor/contributors/develop/coding-guidelines/#strings) recommend using template strings instead of string concatenation whenever possible:
```suggestion
	const namespace = `core/use-filter/${ hookName }`;
```

desaiuditd replied:
d92d8ee

desaiuditd committed:
Prefer template literal over string concatenation.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

youknowriad commented:
While I like the `useFilters` React hook, I find that WP hooks applied to JavaScript are not great in general and tend to avoid them as much as I can. 

An alternative could be to provide this label in the editor settings (we do have other labels there already)

So curious about other thoughts?

desaiuditd commented:
@youknowriad  We briefly touched on it in earlier comments.

As @gziolo said that time, we need to come up with a strategy, naming conventions and probably a starting list of UI labels to put in the filter.

gziolo commented:
`useFilter` is really nice, we should definitely land it 👍

For translations we really need something that allows to set all local overrides on the server so it’s applied only once and then the access to translations is optimized as much as possible. @swissspidy any ideas how we could do it? Maybe a filter on JS call that is added as inline script for each script handle?

youknowriad commented:
Why not a "labels" setting in the block settings object.

gziolo commented:
@youknowriad yes, that’s another way of handling it through the labels object in the settings object 😃 Sweet and simple. It looks like a general purpose `useSetting` React hook would do the trick here or even `useSelect` call.

desaiuditd commented:
@youknowriad @gziolo Can we agree on one approach here? Happy to re-work on this.

talldan commented:
I just happened to see this issue again, and thought about it from a different perspective. The PR description states the following:

> I need to be able to rename the Document label in the sidebar. So that I can differentiate between different post types and custom post types as well. It could also be a generic custom label e.g., Article, Story etc. for any specific newsroom.

When registering a custom post type, labels are already provided. E.g. from the docs:
```php
function wporg_custom_post_type() {
    register_post_type('wporg_product',
        array(
            'labels'      => array(
                'name'          => __('Products', 'textdomain'),
                'singular_name' => __('Product', 'textdomain'),
            ),
                'public'      => true,
                'has_archive' => true,
        )
    );
}
add_action('init', 'wporg_custom_post_type');
```

Maybe the editor should use those labels? It makes sense to me that we call a post a post, or a page a page instead of a document. The only place it might get confusing is for a Block post type.

That would also help solve this issue - https://github.com/WordPress/gutenberg/issues/21850.

gziolo commented:
> Maybe the editor should use those labels? It makes sense to me that we call a post a post, or a page a page instead of a document. The only place it might get confusing is for a Block post type.

I think we have a few places where we pass CPT specific wordings from the server. We should align all of them. I think the idea that Riad shared still applies you just presented where it can be sourced from.

youknowriad commented:
@talldan This is not a bad  idea either, on another issue, someone mentioned the desire to change the "publish" button label too and that also  could be solved with the CPT  labels (that we do use throughout the UI)

mtias commented:
"Document" was always meant to be the actual post type or root element. Originally, a few years ago, we couldn't do it because it wasn't properly exposed. We might still refer to "post" as "document" as an exception because there were some concerns with the fact post also works as a verb that could be confusing in the context of tabs, etc.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd committed:
Use PHP filter instead of JS filter.

desaiuditd commented:
> Why not a "labels" setting in the block settings object.

Hey all @youknowriad @gziolo @talldan , Have re-worked on this approach. It's ready for review again.

desaiuditd commented on lib/client-assets.php:
```diff
@@ -693,3 +693,19 @@ function gutenberg_extend_settings_custom_units( $settings ) {
+ * @return array           Modified editor settings.
+ */
+function gutenberg_extend_settings_ui_labels( $settings ) {
+	$settings['__experimentalLabels'] = array(
```
ℹ️ I had a brief thought of introducing new filter here - just for labels. Something like `apply_filters( 'block_editor_label_setting', $labels )` but then I thought it will be redundant. Vendor plugins can use `block_editor_settings` with lower priority to achieve the same.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

youknowriad commented:
Hey there! sorry for the back and forth, it seems like the latest approach we agreed on and suggested by @talldan is to use the CPT labels instead  of the settings. Here's an example

https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/components/header/fullscreen-mode-close/index.js#L39

desaiuditd commented:
> Hey there! sorry for the back and forth, it seems like the latest approach we agreed on and suggested by @talldan is to use the CPT labels instead of the settings. Here's an example
> 
> https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/components/header/fullscreen-mode-close/index.js#L39

Hey @youknowriad What about @mtias 's comment above?

desaiuditd commented:
I checked the code you referred here - https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/components/header/fullscreen-mode-close/index.js#L39

It seems like, using the post label there make sense. Because it's taking the user back to the List View page for the post type (either default post or custom post type).

Not sure if same applies here, if we give weight to @mtias 's comment.

youknowriad commented:
I think what @mtias is saying is that we should use the post type  label except for the "post" CPT where we should use  "Document".

desaiuditd commented:
Ah. I see. Gotcha.

Let me re-work on that. Thanks for the clarification.

desaiuditd committed:
Use Post Type label for Custom Post Types.

Keep using Document for Posts and Pages.

desaiuditd commented:
@youknowriad Have made the changes.

Have kept `Document` for Posts as well as Pages. Singular Post Type labep for custom post types.

Let me know if Document only needs to be there for Posts and I can change that.

youknowriad commented on lib/client-assets.php:
```diff
@@ -693,3 +693,27 @@ function gutenberg_extend_settings_custom_units( $settings ) {
+
+	return $settings;
+}
+add_filter( 'block_editor_settings', 'gutenberg_extend_settings_ui_labels' );
```
I think this php code should be useless if we rely on the same mechanism (post type) to retrieve the labels used in other components.

desaiuditd replied:
If we don't provide a filter, how do you imagine other plugins can override this value?

My use case is: I have a post meta post category (possible values are Newsletter, Story, Collection, Article etc.) which is not directly related to the core post type. So I need to be able to filter this label value from somewhere.

youknowriad replied:
Curious why you're using a post meta for that and not a separate CPT for each one of these.

desaiuditd replied:
> Curious why you're using a post meta for that and not a separate CPT for each one of these.

Our CMS is closed source for our editorial workflow. So won't be able to give direct example of our architecture. But in summary, we started with two article type Standard and Feature article. Data, input fields and api design for both is same till date. For the most part, it was to only have an affect on the render layer with styles and page layout. Eventually, we ended up having other types such as Newsletters, Live Blog Posts etc. They differ a bit in terms of certain meta fields. But they inherit from the same structure as Standard or Feature Article. So we chose to have a post meta for this. Today, for us, it's mostly about supporting that historical decision in the existing system.

desaiuditd replied:
I'm happy to keep the same source of truth for the post type labels, as long as we have a place to filter it out.

As you said, I can use below as is in SettingsHeader component. But then I may have to add the JS filter which I had added earlier and we moved away from that approach.
```
const { getCurrentPostType } = select( 'core/editor' );
const { getPostType } = select( 'core' );
const postType = getPostType( getCurrentPostType() );
const documentLabel = postType.labels.singular_name
```

youknowriad replied:
I understand your use-case but tbh I'm not sure it warrants an extensibility API and the need to maintain it while the CPT label seems like the most logical choice here.

desaiuditd replied:
@youknowriad I think I will be able to work around this approach which you have suggested. But I will make changes in the PR accordingly. But I'm intrigued by @peterwilsoncc 's suggestion in the comments. It would be worth giving a thought to it.

youknowriad commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -2,22 +2,28 @@
 } ) => {
-	const blockLabel = __( 'Block' );
+	const documentLabel = useSelect( ( select ) => {
+		const settings = select( 'core/block-editor' ).getSettings();
```
As said above, if you check the component  I link previously, you'll see that we already have the info in the post type, we don't  a new setting for that.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

# Conflicts:
#	lib/client-assets.php

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

peterwilsoncc commented:
Considering filters, I think one or the other of these two will need to be available:

* label filtering
* filtering which post types display the generic "Document"

My inclination is for the latter, as this will allow developers/users who prefer a consistent labelling system to do so.

For example:
```php
$show_generic_label = in_array( get_post_type( $post ), [ 'post', 'page' ], true );
$show_generic_label = apply_filters( 'wp_block_editor_generic_document_label', $show_generic_label, $post );
```

desaiuditd commented:
> Considering filters, I think one or the other of these two will need to be available:
> 
> * label filtering
> * filtering which post types display the generic "Document"
> 
> My inclination is for the latter, as this will allow developers/users who prefer a consistent labelling system to do so.
> 
> For example:
> 
> ```
> $show_generic_label = in_array( get_post_type( $post ), [ 'post', 'page' ], true );
> $show_generic_label = apply_filters( 'wp_block_editor_generic_document_label', $show_generic_label, $post );
> ```

@youknowriad Do you have any thoughts on this?

in WP Core, there is huge precedent for customising the text output in UI. What @peterwilsoncc has suggested, I can see that being useful to other developers or people whose clients are using WordPress. So that they can use `Document` for all or no post types as per choice rather than just applying the post type labels.

mcsf commented:
> Considering filters, I think one or the other of these two will need to be available:
> 
>     * label filtering
> 
>     * filtering which post types display the generic "Document"

I'm honestly wondering what use cases you have in mind for this — could you provide some? I think it's important to weigh any possible benefits against not just the cost of expanding the API surface, but also the fact that WP should be able to offer reasonable defaults that one doesn't feel the need to always tweak.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd committed:
Re-use post type labels for Document label.

desaiuditd committed:
Clean up & remove extra code.

desaiuditd committed:
Fix typo.

peterwilsoncc commented:
These are the use cases I had in mind:

**filtering which post types display the generic "Document"**

As I mentioned, this is about consistency: either always displaying the generic label _Document_ or the singular label of the built in or CPT.

In the latter use case, by default the block editor doesn't display the post type anywhere in full screen mode. This allows developers to introduce text to the editor in which users can ensure the content they are about to post is of the correct use type without having to look at the URL.

This is intended to be a low friction API for the Gutenberg team.

**label filtering**

This is much higher friction but allow further flexibility without the need to introduce additional filters in the future. With this change it's overkill.

For tumblr style blogs using Core's post format feature, it would allow developers to display the post format in the UI (for the same reasons as above) based upon the taxonomy used; something along the lines of _Post (status)_ or _Post (gallery)_. Although, in this case updating on the basis of the post state would be more helpful, the labels would still need to be filtered somewhere in the system.

> WP should be able to offer reasonable defaults that one doesn't feel the need to always tweak.

This is certainly the approach the Core contributors take with providing additional options out of the box. I've had plenty of discussion on core.trac.wordpress.org arguing just this.

However, developer filters are very important features for use in client sites and (as @desaiuditd says above, the reason for WP's popularity). 

Without these developer APIs, there'd be no way to extend the HTML allowed for contributors, modify widgets, extend the editors settings: all the features used by the Gutenberg plugin to continue development as a feature plugin. 

This is aside from the SEO improvements that can be made by modifying meta tags, allowing plugins to serve images from a CDN, auto-posting to social media on publication of posts.

mcsf commented:
Thanks for the details, @peterwilsoncc.

> However, developer filters are very important features for use in client sites […]

> Without these developer APIs, there'd be no way to extend the HTML allowed for contributors, modify widgets, extend the editors settings: all the features used by the Gutenberg plugin to continue development as a feature plugin.

I don't mean to diminish the value of developer APIs. Indeed, they are essential to everything we do. :) I just want to be judicious in how we extend the APIs. This is especially true when discussing the possibility of targeting a UI element directly — why not target the underlying data?

Which brings me to the following question: what if we drop "Document" altogether and let the editor always render the name of the current post type's label (`singular_name`)?

This simplifies things a lot in that:

* we avoid arbitrary branching based on conventional post types
* we avoid the pitfalls of targeting a UI element (rendered tab text) instead of the data proper (post type label)
* we can relying on existing and tested APIs ([`post_type_labels_{$post_type}`](https://developer.wordpress.org/reference/hooks/post_type_labels_post_type/)) that sit close to the data in question

I believe that erring on the side of simplicity should benefit both users and developers. CPT label filtering should provide all the flexibility developers may need. Am I missing anything here?

peterwilsoncc commented:
> Which brings me to the following question: what if we drop "Document" altogether and let the editor always render the name of the current post type's label (singular_name)?

Yes, I think that works and makes more sense than yet another filter.

Thanks

ZebulanStanphill commented:
Ditto, it's always seemed a bit odd that we use "Document" there instead of "Post".

mcsf commented:
Thanks for all your comments, it seems we have consensus then! @desaiuditd, are you interested and able to carry on in that direction? Otherwise let me know and I can do it.

desaiuditd commented:
@mcsf Thanks. I was away from my dev machine for awhile. I'll take a look at this today.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd committed:
Use post type label in Document settings tab in sidebar for all post type.

desaiuditd commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -52,14 +69,4 @@ const SettingsHeader = ( {
-			openGeneralSidebar( 'edit-post/block' );
-		},
-	};
-} )( SettingsHeader );
```
This change is not relevant. Just cleaning up to use hooks instead of HoC.

noahtallen replied:
Nice!

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd committed:
Fix e2e tests.

desaiuditd committed:
Fix e2e tests.

desaiuditd committed:
Update docs.

desaiuditd commented:
@mcsf @ZebulanStanphill @youknowriad This is up for review. Some e2e tests were failing due to the new changes. I've fixed them.

Note that this introduces a behaviour change. The document settings tab in the sidebar will always show the post type label for all post types (including Post and Page).

desaiuditd commented:
The PR still shows one pending check `pull-request-automation Expected — Waiting for status to be reported`. I'm not sure what it is waiting for. 🤷‍♂️

mcsf commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -1,46 +1,63 @@
-			: // translators: ARIA label for the Settings Sidebar tab, not selected.
-			  [ __( 'Block' ), '' ];
+			? // translators: ARIA label for the Block Settings Sidebar tab, selected.
+			  [ sprintf( __( '%1$s (selected)' ), blockLabel ), 'is-active' ]
```
Let's avoid the interpolation, I think it contributes to complexity, loss of context, dependency between i18n strings. So let's keep `__( 'Block (selected) ' )`.

If you prefer to avoid the inlined value, you can always define `const blockSelectedLabel` next to `blockLabel`. If so, then I might even suggest turning them into `BLOCK_LABEL` and `BLOCK_SELECTED_LABEL`, to make it clearer that they are placeholders for the actual string.

mcsf commented on packages/e2e-tests/specs/editor/various/preferences.test.js:
```diff
@@ -29,8 +29,10 @@ describe( 'preferences', () => {
+
 		// Open by default.
-		expect( await getActiveSidebarTabText() ).toBe( 'Document' );
+		expect( await getActiveSidebarTabText() ).toBe( documentLabel );
```
If possible, I would always keep the values in these tests explicit. We already know that we are dealing with Posts or Pages, so there is great benefit in reducing the complexity of tests:

```suggestion
		expect( await getActiveSidebarTabText() ).toBe( 'Post' );
```

mcsf commented on packages/e2e-tests/specs/editor/various/sidebar.test.js:
```diff
@@ -43,7 +44,7 @@ describe( 'Sidebar', () => {
 
 		// the active sidebar tab should be document.
-		expect( content ).toBe( ACTIVE_SIDEBAR_BUTTON_TEXT );
+		expect( content ).toBe( documentLabel );
```
See my comment in `preferences.test.js`.

mcsf commented on packages/e2e-tests/specs/editor/various/keyboard-navigable-blocks.test.js:
```diff
@@ -187,7 +188,8 @@ describe( 'Order of block keyboard navigation', () => {
 		await page.keyboard.press( 'Tab' );
-		await expect( await getActiveLabel() ).toBe( 'Document' );
+		const documentLabel = await getDocumentLabel();
+		await expect( await getActiveLabel() ).toBe( documentLabel );
```
Same feedback as in the other suites. :)

mcsf commented on packages/e2e-test-utils/src/post-type.js:
```diff
@@ -0,0 +1,23 @@
+ */
+import { wpDataSelect } from './wp-data-select';
+
+/**
+ * Returns a promise which resolves with the current post type name.
+ *
+ * @return {Promise<?string>} Promise resolving with current post type name.
+ */
+export async function getCurrentPostType() {
+	return wpDataSelect( 'core/editor', 'getCurrentPostType' );
+}
+
+/**
+ * Returns a promise which resolves with the post type object for the given post type name.
+ *
+ * @param  {string}           name Name of the post type.
+ * @return {Promise<?Object>}      Promise resolving with post type object.
+ */
+export async function getPostType( name ) {
+	return wpDataSelect( 'core', 'getPostType', name );
+}
```
There are two important things here:

* In E2E tests we strongly discourage tapping into the data layer for reads and effects. This is mainly because the point of E2E tests is to simulate the experience of users, and so it should be possible to trigger change and observe results via the editor UI — even if that feels more fragile to us as developers, it ensures more faithful tests.

* Even if/when it seems unavoidable that a given test would benefit from such a function, we need to be careful about adding helpers to `e2e-test-utils`, because from that point on they become public interfaces that we will have to maintain.

Specifically, in the context of the tests of this PR, I would say:

* We don't need to read the post-type labels, and so we don't need these selector-like helpers.
* Even if we did, such logic would have to live in the file of the respective test suite.

mcsf commented on packages/e2e-test-utils/src/document-label.js:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @return {string} Label string for Document settings tab in sidebar.
+ */
+export async function getDocumentLabel() {
+	const currentPostType = await getCurrentPostType();
+	const postType = await getPostType( currentPostType );
+	return get( postType, [ 'labels', 'singular_name' ], 'Document' );
+}
```
Same feedback as for `post-type.js`.

mcsf reviewed (changes requested):
Thanks for pivoting! This looks good, but I flagged some stuff that needs fixing.

mcsf commented:
> The PR still shows one pending check `pull-request-automation Expected — Waiting for status to be reported`. I'm not sure what it is waiting for. 🤷‍♂️

This seems to have resolved in the meantime, no?

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd committed:
Remove complexities around e2e tests.

desaiuditd committed:
Clean up.

desaiuditd committed:
Clean up.

desaiuditd commented:
> Thanks for pivoting! This looks good, but I flagged some stuff that needs fixing.

@mcsf I've addressed the feedback. This should be again good to go.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd commented:
@WordPress/gutenberg Can someone help to review this PR?

noahtallen commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -1,65 +1,73 @@
+// translators: Label for the Block Settings Sidebar tab, not selected.
+const BLOCK_LABEL = __( 'Block' );
+// translators: Label for the Block Settings Sidebar tab, selected.
+const BLOCK_SELECTED_LABEL = __( 'Block (selected)' );
```
Why not keep these inline?

desaiuditd replied:
The strings were appearing at multiple places. Hence, I made it into a constant to re-use across the file. Following Miguel's suggestion.

mcsf replied:
Ref to my comment: https://github.com/WordPress/gutenberg/pull/17311#discussion_r447819934

My main concern was the string interpolation, that was the one thing that needed to be fixed. As for reusing the string or inlining it, I'd say that's up to you (my personal preference is probably inlining).

desaiuditd replied:
I can put it back inline.

Side question out of curiosity: If I add same string again in the same file (or maybe anywhere else), do I have to add the `translators: ...` comment again?

desaiuditd replied:
b3a060e

noahtallen replied:
Honestly I am not sure about the translators comment :)

noahtallen commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -1,65 +1,73 @@
+		// translators: Default label for the Document sidebar tab, not selected.
+		const defaultDocumentLabel = __( 'Document' );
+
+		return get(
```
Could use the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) to avoid pulling the lodash dep:

```js
return postType?.labels?.singular_name ?? __( 'Document' );

desaiuditd replied:
Unfortunately, I can't.

`postType?.labels?.singular_name` gives me eslint error. `Identifier 'singular_name' is not in camel case.eslintcamelcase`

The eslint error goes away if I use `postType.labels.singular_name` without the optional chaining. But then it defeats the purpose. Okay, if I keep it as is?

swissspidy replied:
That's really a false positive in this case. Just disable it :-)

```js
// Disable reason: Post type labels object is shaped like this.
// eslint-disable-next-line name-of-the-eslint-rule
return postType?.labels?.singular_name ?? __( 'Document' );
```

noahtallen commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -1,65 +1,73 @@
 			<li>
 				<Button
-					onClick={ openDocumentSettings }
+					onClick={ () => openGeneralSidebar( 'edit-post/document' ) }
```
I'm not sure what our standards are for creating these functions inline, but it might be worth just bumping these to the main body of the component like:

```js
const openDocumentSettings = () => openGeneralSidebar( 'edit-post/document' );
const openBlockSettings = () => openGeneralSidebar( 'edit-post/block' );
```

mcsf replied:
> I'm not sure what our standards

There are no standards for this, but inlining the function is very common nowadays. I would keep as it was, since it keeps everything that matters in one place, especially with such simple functions (they're basically a `Function#bind` call with a different syntax). But it's not a rule, either way would be fine.

Related reading: [React, Inline Functions, and Performance](https://reacttraining.com/blog/react-inline-functions-and-performance/) by Ryan Florence.

swissspidy replied:
Leaving performance out of the picture here, keeping

```js
const openDocumentSettings = () => openGeneralSidebar( 'edit-post/document' );
const openBlockSettings = () => openGeneralSidebar( 'edit-post/block' );
```

seems a bit more readable IMO :)

mcsf replied:
It's up to y'all. :)

desaiuditd replied:
14f0f0a

noahtallen replied:
Thanks for the link! Looks like an interesting read :)

noahtallen reviewed (commented):
Just left some mostly nit-picky comments about code :) 

Overall, this change looks pretty good to me. I just haven't gotten a chance to test it yet.

swissspidy commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -1,65 +1,73 @@
-			  [ __( 'Document (selected)' ), 'is-active' ]
-			: // translators: ARIA label for the Document sidebar tab, not selected.
-			  [ __( 'Document' ), '' ];
+			  [ sprintf( __( '%1$s (selected)' ), documentLabel ), 'is-active' ]
```
```suggestion
			  [ sprintf( __( '%s (selected)' ), documentLabel ), 'is-active' ]
```

swissspidy commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -1,65 +1,73 @@
-	const blockLabel = __( 'Block' );
 	const [ documentAriaLabel, documentActiveClass ] =
 		sidebarName === 'edit-post/document'
 			? // translators: ARIA label for the Document sidebar tab, selected.
```
```suggestion
			? // translators: ARIA label for the Document sidebar tab, selected. %s: Document label.
```

swissspidy commented:
@desaiuditd What does this look like in the sidebar when editing reusable blocks?

mcsf reviewed (commented):
Nice work. We just need to address the feedback from Noah and Pascal, and then we'll be good to go.

desaiuditd commented:
> @desaiuditd What does this look like in the sidebar when editing reusable blocks?

@swissspidy Good catch. It reads `Block` for both the tabs.

![image](https://user-images.githubusercontent.com/2236554/88181242-e98e2880-cc71-11ea-9b13-cbf7b60387c1.png)

How do we want to handle this?

desaiuditd committed:
Update packages/edit-post/src/components/sidebar/settings-header/index.js

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

desaiuditd committed:
Update packages/edit-post/src/components/sidebar/settings-header/index.js

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

desaiuditd committed:
Fix nits.

desaiuditd committed:
Fix nits.

desaiuditd committed:
Fix nits.

ZebulanStanphill commented:
@desaiuditd It seems to me that the Reusable Block post type label just needs to be changed to "Reusable Block".

afercia commented:
Are we sure we want to expose the UI to the risk of custom post types that could have a very, very, long name especially in some languages? We can't predict a CPT name length and CSS techniques like `text-overflow: ellipsis;` don't seem a good idea to me 🙂 

<img width="358" alt="Screenshot 2020-07-22 at 17 46 56" src="https://user-images.githubusercontent.com/1682452/88198162-b5037800-cc43-11ea-92aa-98070d270f2e.png">

mcsf commented:
> Are we sure we want to expose the UI to the risk of custom post types that could have a very, very, long name especially in some languages?

I think that's a perfectly acceptable risk. Labels are, after all, "official" user-facing names of a CPT. They are arguably expected to be succinct. For instance, the generic label `name` is the fallback for `menu_name`, which is a label that needs to fit in WP-Admin's sidebar.

Case in point: the label "A very long Custom Post Type name" still fits, even if not elegantly. It becomes a problem at three lines.

> We can't predict a CPT name length and CSS techniques like `text-overflow: ellipsis;` don't seem a good idea to me

IMO this is the kind of issue that we'd only look at if and when it proves to be an issue.

mapk commented:
Related: https://github.com/WordPress/gutenberg/issues/20877#issuecomment-661934064 That particular exploration from @dubielzyk brings the Document tab into the menu popover.

desaiuditd committed:
Merge branch 'master' into add/filter-to-override-document-label-in-settings-header

desaiuditd committed:
Add shim to override post type labels for Reusable blocks.

desaiuditd committed:
Adds trac link.

desaiuditd committed:
Fix i18n.

desaiuditd commented:
@swissspidy @ZebulanStanphill @mcsf I've added the shim to override the post type labels for reusable blocks.

Respective trac ticket to update it in WordPress Core [Trac#50755](https://core.trac.wordpress.org/ticket/50755)

![image](https://user-images.githubusercontent.com/2236554/88395979-3a825600-ce05-11ea-8d6f-fdc4584672ae.png)

mcsf reviewed (approved):
Looks great, thanks for all the work.
